File: daemon/src/routes/diff.ts
import { basis, query, root } from './request'
import type { RouteTable } from './route'

/** Two branches compared whole. Nothing here is about a commit: what is reported is the
 *  difference between the branch you are on and the branch you named — as the two stand,
 *  or against where they parted, which is what the basis says. */
export const diff = {
  'GET /api/diff': async (c, ctx) =>
    c.json({ files: await ctx.branches.diff(root(c), query(c, 'against'), basis(c)) }),

  'GET /api/diff/file': async (c, ctx) =>
    c.json(await ctx.branches.file(root(c), query(c, 'against'), query(c, 'path'), basis(c))),
} satisfies RouteTable

File: daemon/internal/api/diff.go
// Two branches compared whole.

package api

import (
	"net/http"

	"github.com/broodmotherai/broodmother/daemon/internal/app"
	"github.com/broodmotherai/broodmother/daemon/internal/doc"
	"github.com/broodmotherai/broodmother/daemon/internal/git"
)

// basisOf: anything but the word `split` is `now`, the way the TypeScript reads it — a basis is
// a choice between two pictures rather than a field to be refused over.
func basisOf(r *http.Request) git.Basis {
	if r.URL.Query().Get("basis") == string(git.Split) {
		return git.Split
	}
	return git.Now
}

var diffTable = Table{
	"GET /api/diff": func(_ http.ResponseWriter, r *http.Request, ctx *app.Context) (any, error) {
		root, against, err := comparing(r)
		if err != nil {
			return nil, err
		}
		files, err := ctx.Branches.Diff(root, against, basisOf(r))
		if err != nil {
			return nil, err
		}
		return map[string]any{"files": files}, nil
	},

	"GET /api/diff/file": func(_ http.ResponseWriter, r *http.Request, ctx *app.Context) (any, error) {
		root, against, err := comparing(r)
		if err != nil {
			return nil, err
		}
		path, err := query(r, "path")
		if err != nil {
			return nil, err
		}
		return ctx.Branches.DiffFile(root, against, doc.Path(path), basisOf(r))
	},
}

// comparing is the root a diff is asked of and the branch it is held against.
func comparing(r *http.Request) (doc.Root, string, error) {
	root, err := rootOf(r)
	if err != nil {
		return "", "", err
	}
	against, err := query(r, "against")
	if err != nil {
		return "", "", err
	}
	return root, against, nil
}
